### PR TITLE
add transformErrorReponse to rtkq endpoints

### LINKED ISSUE #1139: RTKQ feature proposal: Add transformError property to the endpoints
It would be great to have a `transformResponse` equivalent for errors. The `transformError` property would allow manipulation of the error returned by a query or mutation before it hits the cache.

This new property would help to decouple error parsing logic from the view. The service will encapsulate the logic, and the views will act just as a presenter of the data/error the query or mutation returns.

### PR DESCRIPTION
**Pull request by dreyks**
Closes #1139

The first draft of the PR to get early feedback to know if I'm doing the right thing

I'll add more tests and doc later

### CODE REVIEW

**Comment by dreyks** [comment]
hey! I'm sorry for such a huge delay. I'll get this into shape soon

File: packages/toolkit/src/query/core/buildThunks.ts [hunk 1]
```diff
@@ -344,8 +344,25 @@ export function buildThunks<
...
+        if (endpointDefinition.query && endpointDefinition.transformErrorResponse) {
+          transformErrorResponse = endpointDefinition.transformErrorResponse
+        }
+        try {
+          return rejectWithValue(
+              await transformErrorResponse(catchedError.value, catchedError.meta, arg.originalArgs),
+              { baseQueryMeta: catchedError.meta }
+          )
+        } catch (e) {
+          catchedError = e
```
Comment (dreyks) [inline packages/toolkit/src/query/core/buildThunks.ts:364]: is it an ok thing to do? didn't want to repeat the whole console.error/throw block here

File: packages/toolkit/src/query/tests/queryFn.test.tsx [hunk 2]
```diff
@@ -9,9 +9,9 @@ import type { QuerySubState } from '@reduxjs/toolkit/dist/query/core/apiState'
 
 describe('queryFn base implementation tests', () => {
   const baseQuery: BaseQueryFn<string, { wrappedByBaseQuery: string }, string> =
-    jest.fn((arg: string) => ({
-      data: { wrappedByBaseQuery: arg },
-    }))
+    jest.fn((arg: string) => arg.includes('withErrorQuery')
+      ? ({ error: `cut${arg}` })
+      : ({ data: { wrappedByBaseQuery: arg } }))
```
Comment (dreyks) [inline packages/toolkit/src/query/tests/queryFn.test.tsx:14]: something is happening here that i do not understand

I've tried fully mimicking the data case like this
```ts
const baseQuery: BaseQueryFn<string, { wrappedByBaseQuery: string }, { wrappedByBaseQuery: string }> =
    jest.fn((arg: string) => arg.includes('withErrorQuery')
      ? ({ error: { wrappedByBaseQuery: arg } })
      : ({ data: { wrappedByBaseQuery: arg } }))
```

but this causes cases like `withErrorQueryFn`
```ts
withErrorQueryFn: build.query<string, string>({
  queryFn(arg: string) {
    return { error: `resultFrom(${arg})` }
  },
}),
```
 to fail the typecheck
> Type '(arg: string) => { error: string; }' is not assignable to type '(arg: string, api: BaseQueryApi, extraOptions: {}, baseQuery: (arg: string) => MaybePromise<QueryReturnValue<{ wrappedByBaseQuery: string; }, { wrappedByBaseQuery: string; }, {}>>) => MaybePromise<...>'. 
  Type '{ error: string; }' is not assignable to type 'MaybePromise<QueryReturnValue<string, { wrappedByBaseQuery: string; }, unknown>>'. 
    Property 'data' is missing in type '{ error: string; }' but required in type '{ error?: undefined; data: string; meta?: unknown; }'.

i lack proper TS knowledge here so I just went with not changing the `string` error response type. But I'll be glad to do it better if someone hints me on what might be the issue

**Review by dreyks — COMMENTED** [review]
rebased, added the docs, and more tests. ready for review

**Comment by markerikson** [comment]
Okay, fixed up some TS issues in the `devWarning` test, and I think this is ready now. Thank you for putting this together!

**Pull request merged**